ABA-127: Attendee Statistics in Event Administrate

AdrianAndersen opened the pull request:
In this PR, I have implemented visualisations of attendees for each event in. /administrate/. The following statistics are displayed:
- Gender distribution (male, female, other)
- Datatek class distribution
- Komtek class distribution
- Datatek vs Komtek (+ PhD and graduated people) distribution
- Group affiliation (Abakom, Revy or Abakus)
- Registration/Unregistrations per day [This might be useful for seeing how effective "blesting" for a given event is]

Hopefully this functionality can help event organisers be more data driven when planning and executing events. 

## Screenshot of the graphs from from this years Charity Party (Veldedighetsfest)
![image](https://user-images.githubusercontent.com/26925695/199369225-c25548cb-72c7-4b40-8cc0-89119fd16117.png)![image](https://user-images.githubusercontent.com/26925695/199369249-5cfa3b1c-d9d1-43dc-a89e-4115b2cac509.png)

eikhr commented:
I think it's a great idea! I can imagine that f.ex. the distribution of what year the attendees are going could be very useful for somone arranging a course. I think we have a lot of very useful data that is currently too diffucult to access, so I'm all for making this more easily available.

It could also be interesting to get the share of attendees who are comitee (and maybe revue) members, although this might be a bit more difficult to fetch data for.

ollfkaih commented:
Good idea, looks good overall. Couple of nitpicks:  
Localize the labels for gender.
Also, maybe we should hide numbers below a certain percentage or show the outside the pie, if that is possible?

eikhr commented on app/routes/events/components/EventAttendeeStatistics.js:
```diff
@@ -0,0 +1,123 @@
+  distributionData: DistributionDataPoint[],
+}) => {
+  return (
+    <Flex alignItems="center">
+      <PieChart width={400} height={350}>
+        <Pie
+          data={distributionData}
+          cx={200}
+          cy={150}
+          labelLine={false}
+          label={renderCustomizedLabel}
+          dataKey="count"
+        >
+          {distributionData.map((entry, index) => (
+            <Cell
+              key={`cell-${index}`}
+              fill={CHART_COLORS[index % CHART_COLORS.length]}
+            />
+          ))}
+        </Pie>
+      </PieChart>
+      <div>
+        <ul>
+          {distributionData.map((dataPoint, i) => (
+            <li key={i}>
+              <span
+                style={{
+                  backgroundColor: CHART_COLORS[i % CHART_COLORS.length],
+                }}
+              >
+                &nbsp;
+              </span>
+              <span> {dataPoint.name}</span>
+            </li>
+          ))}
+        </ul>
+      </div>
+    </Flex>
+  );
+};
```
I love that the pie chart is a separate component and abstract enough to be used for other things as well! 

I feel like it should be moved into a separate file in /components, and possibly also be reused in the survey result page, as that chart seems to be very similar.

eikhr replied:
Actually, while you're at it you should probably move the bar chart from survey results into a separate component as well. No reason to have all this chart-creation code inside the survey route folder if we also use charts outside of surveys.

AdrianAndersen replied:
Yeah, I agree! ✅

AdrianAndersen commented:
Thank you so much for the initial discussion and feedback for the usability of this feature! 😄 

I have now implemented your suggestions and refactored out some of the code from the surveys to make it more modular. Hope you have time to take a look at the changes. (Even though it turned out the be quite a few lines of code 👨🏻‍🎨 )

AdrianAndersen commented on app/routes/events/components/EventAttendeeStatistics.js:
```diff
@@ -0,0 +1,283 @@
+  unregistrations: number;
+}
+
+const ABAKOM_GROUP_IDS = [4, 5, 6, 7, 8, 9, 10, 11, 98, 231];
+const REVY_GROUP_IDS = [59, 119, 120, 121, 123, 124, 153, 154, 214, 222, 225];
```
It is kinda ugly to hard code the group IDs here, since it can get outdated if we for instance add another Abakom/Revy group. if someone has any suggestions for how to improve this, please let me know. 👀

eikhr replied:
Yeah, we do have `GroupTypeCommitee` and `GroupTypeRevue`, but group type isn't fetched for users in registrations so you'd have to make some backend changes, or do some extra API-calls.

LudvigHz replied:
Yeah, I think the most sensible here if you want it, is to fetch `/api/v1/groups/?type=<komite/revy>` and use the data from that to find what group type people are in.

erlingfn replied:
I agree with Ludvig. If we add more committees I bet we will not remember to add them here

AdrianAndersen replied:
Yeah, that sounds like a good solution. Moved it now ✅

LudvigHz commented:
As far as I understood, this is currently on the `/administrate/attendees` route?

If so, I think we need to move this to another route, f.ex. `/administrate/statistics`. This is so we don't make the page _enourmous_ and can split out the data fetching.

AdrianAndersen commented:
> As far as I understood, this is currently on the /administrate/attendees route?

Yup, currently it is. It makes sense to move it to a separate page, I fix.

ingraso commented on app/routes/events/components/EventAttendeeStatistics.js:
```diff
@@ -0,0 +1,283 @@
+    count: 0,
+  };
+  const komTekTotal: DistributionDataPoint = {
+    name: 'Komunikasjonsteknologi',
```
As the name of the study has changed I suggest also updating it here, in addition to adding an "m".
```suggestion
    name: 'Kommunikasjonsteknologi og digital sikkerhet',
```

AdrianAndersen replied:
✅

erlingfn commented on app/routes/events/components/EventAttendeeStatistics.js:
```diff
@@ -0,0 +1,283 @@
+  return 'Annet';
+};
+
+const createGenericDataPoint = (
```
Perhaps a better name for this might be either `addGenericDataPoint` or `addOrIncrementDataPoint`? I was a bit confused by the naming :grimacing:

AdrianAndersen replied:
✅

erlingfn commented on app/routes/events/components/EventAttendeeStatistics.js:
```diff
@@ -0,0 +1,283 @@
+  }
+};
+
+const createGroupDataPoint = (
```
Same as above applies to this.

AdrianAndersen replied:
✅

erlingfn commented on app/components/Chart/utils.js:
```diff
@@ -0,0 +1,60 @@
+  count: number;
+}
+
+export const CHART_COLORS = [
```
Would it be possible to include more chart colours? I have sometimes experienced charts being difficult to read because it ran out of colours.

AdrianAndersen replied:
Yup, I added some more colors now, so that we now have 12 colors. That should suffice for all charts. ✅

erlingfn reviewed (approved):
Great work! A few nitpicks, but they may just be me.

AdrianAndersen commented:
Also, kinda weird that `tsc` and some seemingly unrelated Cypress tests fails in CI. 👀

erlingfn commented on app/routes/events/EventStatisticsRoute.ts:
```diff
@@ -0,0 +1,43 @@
+import { selectGroupsWithType } from 'app/reducers/groups';
+import { GroupTypeCommittee, GroupTypeRevue } from 'app/models';
+
+const groups: {
+  committee: typeof GroupTypeCommittee;
+  revue: typeof GroupTypeRevue;
+} = {
+  committee: 'komite',
+  revue: 'revy',
+};
```
Are they in the redux state at all? I do not know if they are in the serializer.

LudvigHz replied:
Hmm, I tried using them directly and it worked just fine :thinking: 

Ideally though, if you would be a champ @AdrianAndersen, you can just yeet all the `GroupType****` stuff from `app/models` and replace it with an enum, which can be used both as a type and value everywhere :)

So something like this in app/models:
```ts
export enum GroupType {
  committee = 'komite',
  Board = 'styre',
  Revue = 'revy',
  Interest = 'interesse',
  Grade = 'klasse',
  SubGroup = 'under',
  Other = 'annen',
}
```
And use that in the route.

It's not a HUGE deal though, I can add a task for converting every string type in app/models to enums, but it's nice to just do it while you're working with it :)

AdrianAndersen replied:
Huh, interesting. It actually worked to access the variables directly today. I am guessing it was just a hot reloading issue I experienced earlier. 🤔 

I refactored all GroupType** usages to an enum now. ✅

eikhr commented:
I just noticed that looking at stats for old events might be a bit misleading. It will show the status of each attendee now, not when the event took place. This means that an event that happened two years ago will show that no 1st or 2nd graders attended. It also makes it look like more non-committee/revue members attended than you would expect, as all the people who are now done studying are no loger counted as committee/revue-members.

I can't think of any easy way to fix this, but perhaps a warning on the statistics page for old events would be a good idea.

AdrianAndersen commented:
> I can't think of any easy way to fix this, but perhaps a warning on the statistics page for old events would be a good idea.

@eikhr, I guess one could fetch a users previous memberships, but I think that is too cumbersome. I therefore added a warning on events from a previous semester as you suggested:

<img width="1041" alt="image" src="https://user-images.githubusercontent.com/26925695/200434389-ca4db076-3294-4a87-9cb5-6de28073150d.png">

LudvigHz reviewed (approved):
LGTM :shipit:
